servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -143,37 +144,43 @@ public void onSubscribe(final Subscription s) {
             }
 
-            @SuppressWarnings("unchecked")
             @Override
             public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                 LOGGER.debug("Load balancer {}, received new ServiceDiscoverer event {}.", RoundRobinLoadBalancer.this,
                         event);
+                @SuppressWarnings("unchecked")
                 final List<Host<ResolvedAddress, C>> activeAddresses =
-                        activeHostsUpdater.updateAndGet(RoundRobinLoadBalancer.this, currentAddresses -> {
-                            final List<Host<ResolvedAddress, C>> refreshedAddresses =
-                                    new ArrayList<Host<ResolvedAddress, C>>(currentAddresses);
-                            final ResolvedAddress addr = requireNonNull(event.address());
-                            for (int i = 0; i < refreshedAddresses.size(); i++) {
-                                Host<ResolvedAddress, C> host = refreshedAddresses.get(i);
+                    activeHostsUpdater.updateAndGet(RoundRobinLoadBalancer.this, oldHosts -> {
+                        final ResolvedAddress addr = requireNonNull(event.address());
+                        @SuppressWarnings("unchecked")
+                        final List<Host<ResolvedAddress, C>> oldHostsTyped = (List<Host<ResolvedAddress, C>>) oldHosts;
+                        if (event.isAvailable()) {
+                            if (oldHostsTyped.isEmpty()) {
+                                return singletonList(new Host<>(addr));
+                            }
+                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() + 1);
+                            newHosts.addAll(oldHostsTyped);
+                            newHosts.add(new Host<>(addr));
+                            return newHosts;
+                        } else if (oldHostsTyped.isEmpty()) {
+                            return emptyList();
+                        } else {
+                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() - 1);
+                            for (int i = 0; i < oldHostsTyped.size(); ++i) {
+                                final Host<ResolvedAddress, C> host = oldHostsTyped.get(i);
                                 if (host.address.equals(addr)) {
-                                    if (event.isAvailable()) {
-                                        LOGGER.debug("Address {} added but it already exists.", addr);
-                                        return currentAddresses;
-                                    } else {
-                                        refreshedAddresses.remove(i);
-                                        host.markInactive();
-                                        return refreshedAddresses;
+                                    host.markInactive();
+                                    for (int x = i + 1; x < oldHostsTyped.size(); ++x) {
```

Comment:
We said that it's ok to have duplicated addresses from SD. Neither SD, nor RRLB removes them. What is expected behavior when we receive inactive event for an address that appears multiple times in the list? Should we remove all entries or only the first one?

Reply:
initially this code was removing all items that matched, however I updated it last night to update just the first match. for each available event there should be a corresponding unavailable event. if duplicates are used for weighting (e.g. duplicate entries make it more likely to select) then adjusting the weighting by removing an entry shouldn't remove all the entries.